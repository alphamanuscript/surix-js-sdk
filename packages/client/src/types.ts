import { WrappedEntity, WrappedFile } from '@surix/data-helpers';

export interface Project {
  entities: ProjectEntities;
  files: ProjectFiles;
  tags: ProjectTags;
}

export interface ProjectEntities {
  get(id: string): Promise<WrappedEntity>;
  query(query?: Query): Promise<WrappedEntity[]>;
  create(entity: Entity): Promise<WrappedEntity>;
  put(entity: Entity): Promise<WrappedEntity>;
}

export interface ProjectFiles {
  get(id: string): Promise<WrappedFile>;
  list(): Promise<WrappedFile[]>;
}

export interface ProjectTags {
  list(): Promise<TagList>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];

export enum FieldType {
  TEXT = 'text',
  BOOLEAN = 'boolean',
  NUMBER = 'number',
  DATETIME = 'datetime',
  PHONE = 'phone',
  EMAIL = 'email',
  FILE = 'file',
  OBJECT = 'object',
  LIST = 'list'
}

export interface FieldPair {
  type: FieldType;
  value: RawEntityData | string | RawEntityData[] | number;
}

export interface RawEntityField extends FieldPair {
}

export interface RawEntityData {
  [fieldName: string]: RawEntityField;
}

/**
 * this interface describe the shape of the entity
 * as used within the system and the API
 */
export interface RawEntity {
  data: RawEntityData;
  tags?: string[];
}

export interface DataField {
  [fieldName: string]: string | number | object | boolean;
}
export interface Entity {
  data: DataField;
  tags?: string[];
  _id?: string;
}