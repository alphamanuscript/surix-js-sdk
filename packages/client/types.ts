import { EntityWrapper } from '@surix/data-helpers';

export interface Project {
  db: ProjectDatabase;
  files: ProjectFiles;
}

export interface ProjectDatabase {
  get(id: string): Promise<EntityWrapper>;
  query(query: Query): Promise<EntityWrapper[]>;
  getTags(): Promise<TagList[]>;
}

export interface ProjectFiles {
  get(id: string): Promise<ApiFile>;
  list(): Promise<ApiFile[]>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface ApiFile {
  name: string;
  status: 'failed' | 'ready' | 'pending';
  public: boolean;
  downloadUrl: string;
  tags: TagList[];
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];