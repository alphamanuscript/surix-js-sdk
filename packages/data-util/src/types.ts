export class ApiEntity {
    _id: string;
    createdAt: string;
    updatedAt: string;
    data: ObjectValue;
    createdBy: AuthContext;
    tags: string[];
}

export type FieldType = 'text' | 'boolean' | 'number' | 'datetime'
    | 'file' | 'object' | 'list';

export type FieldValue = string | boolean | ListItemValue[] | ObjectValue;

export interface ListItemValue {
    type: FieldType;
    value: FieldValue;
}

export interface ObjectValue {
    [fieldName: string]: {
        name: string;
        type: FieldType;
        label: string;
        value: FieldValue;
    };
}

export interface AuthContext {
    _id: string;
    type: 'user' | 'app' | 'localapp';
}