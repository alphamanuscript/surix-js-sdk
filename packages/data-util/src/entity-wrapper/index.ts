export { EntityWrapper } from './entity-wrapper';
