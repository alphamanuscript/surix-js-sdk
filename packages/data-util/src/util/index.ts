export {
  dehydrateValue,
  makeField,
  toFieldPathArray,
  walkEntityPath,
  walkObjectPath
} from './util';