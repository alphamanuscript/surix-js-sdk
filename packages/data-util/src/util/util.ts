import {
  ApiEntity,
  Field,
  FieldPathArray,
  FieldType,
  FieldValue,
  ListValue,
  ObjectValue } from '../types';

export function toFieldPathArray (fieldPath: string): FieldPathArray {
  return fieldPath.split('.');
}

export function walkEntityPath (entity: ApiEntity, path: FieldPathArray): any {
  return walkObjectPath(entity.data, path);
}

export function walkObjectPath (obj: ObjectValue | ListValue, path: FieldPathArray): any {
  const [firstKey, ...restPath] = path;
  if (typeof firstKey === 'undefined') {
    return Array.isArray(obj) ? dehydrateList(obj) : dehydrateObject(obj);
  }
  if (firstKey in obj) {
    const field = obj[firstKey];
    if (field.type === 'object' || field.type === 'list') {
      return walkObjectPath(field.value, restPath);
    }
    return dehydrateValue(field.value, field.type);
  }
}

export function dehydrateValue (value: any, type: FieldType): any {
  switch (type) {
    case 'datetime':
      return new Date(value);
    case 'number':
      return Number(value);
    case 'list':
      return dehydrateList(value);
    case 'object':
      return dehydrateObject(value);
    case 'boolean':
    case 'text':
    case 'file':
    default:
      return value;
  }
}

export function dehydrateText (textValue: string): string {
  return textValue;
}

export function dehydrateNumber (numValue: string): number {
  return Number(numValue);
}

export function dehydrateObject (obj: ObjectValue): object {
  return Object.keys(obj)
    .reduce((result, key) => {
      const { type, value } = obj[key];
      return {
        ...result,
        [key]: dehydrateValue(value, type)
      };
    }, {});
}

export function dehydrateList (list: ListValue): any[] {
  return list.map(({ type, value }) => dehydrateValue(value, type));
}

export function makeField (value: FieldValue, type: FieldType, name: string, label: string): Field {
  return {
    value,
    type,
    name,
    label
  };
}