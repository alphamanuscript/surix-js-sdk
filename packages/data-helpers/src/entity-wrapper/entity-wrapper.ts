import {
  ApiEntity,
  AuthContext,
  Field,
  FieldOrValuePair,
  FieldPathArray,
  FieldType,
  NullableFieldOrValuePair,
  WrappedEntity } from '../types';
import { deflateValue, toFieldPathArray, walkEntityPath } from '../util';

const normalizeFieldPath = (fieldPath: string|FieldPathArray): FieldPathArray =>
   Array.isArray(fieldPath) && fieldPath || toFieldPathArray(fieldPath);

export class EntityWrapper implements WrappedEntity {
  private _entity: ApiEntity;
  private _dataCache: object;

  createdAt: Date;

  updatedAt: Date;

  constructor (rawEntity: ApiEntity) {
    this._entity = rawEntity;
    this.createdAt = new Date(this._entity.createdAt);
    this.updatedAt = new Date(this._entity.updatedAt);
  }

  get rawEntity (): ApiEntity {
    return this._entity;
  }

  get _id (): string {
    return this._entity._id;
  }

  get id (): string {
    return this._id;
  }

  get tags (): string[] {
    return this._entity.tags;
  }

  get createdBy (): AuthContext {
    return this._entity.createdBy;
  }

  get (fieldPath: string | FieldPathArray, defaultValue?: any): any {
    const pathArray = normalizeFieldPath(fieldPath);
    return this._value(pathArray, defaultValue);
  }

  private _value (fieldPath: FieldPathArray, defaultValue?: any): any {
    const field = walkEntityPath(this._entity, fieldPath);
    return field ? deflateValue(field.value, field.type) : defaultValue;
  }

  type (fieldPath: string | FieldPathArray, defaultType?: FieldType): FieldType|undefined {
    const pathArray = normalizeFieldPath(fieldPath);
    const field = walkEntityPath(this._entity, pathArray);
    return field ? field.type : defaultType;
  }

  label (fieldPath: string | FieldPathArray, defaultLabel?: string): string|undefined {
    const pathArray = normalizeFieldPath(fieldPath);
    const field = walkEntityPath(this._entity, pathArray) as Field;
    return field && field.label ? field.label : defaultLabel;
  } 

  field (fieldPath: string | FieldPathArray, defaultField?: FieldOrValuePair):
  NullableFieldOrValuePair {
    const pathArray = normalizeFieldPath(fieldPath);
    const field = walkEntityPath(this._entity, pathArray);
    return field || defaultField;
  } 

  data (): object {
    if (!this._dataCache) {
      this._dataCache = deflateValue(this._entity.data, 'object');
    }
    return this._dataCache;
  }
}

export function wrapEntity (entity: ApiEntity): WrappedEntity {
  return new EntityWrapper(entity);
}

export function wrapEntityArray (entities: ApiEntity[]): WrappedEntity[] {
  return entities.map(wrapEntity);
}