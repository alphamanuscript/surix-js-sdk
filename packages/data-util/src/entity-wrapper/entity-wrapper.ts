import { ApiEntity, AuthContext, FieldPathArray } from '../types';
import { dehydrateValue, toFieldPathArray, walkEntityPath } from '../util';

export class EntityWrapper {
  private _entity: ApiEntity;
  constructor (rawEntity: ApiEntity) {
    this._entity = rawEntity;
  }

  get rawEntity (): ApiEntity {
    return this._entity;
  }

  get _id (): string {
    return this._entity._id;
  }

  get tags (): string[] {
    return this._entity.tags;
  }

  get createdBy (): AuthContext {
    return this._entity.createdBy;
  }

  get createdAt (): Date {
    return new Date(this._entity.createdAt);
  }

  get updatedAt (): Date {
    return new Date(this._entity.updatedAt);
  }

  value (fieldPath: string | FieldPathArray): any {
    const pathArray = Array.isArray(fieldPath) && fieldPath || toFieldPathArray(fieldPath);
    return this._value(pathArray);
  }
  _value (fieldPath: FieldPathArray): any {
    const field = walkEntityPath(this._entity, fieldPath);
    if (field) {
      return dehydrateValue(field.value, field.type);
    }
  }
    
}