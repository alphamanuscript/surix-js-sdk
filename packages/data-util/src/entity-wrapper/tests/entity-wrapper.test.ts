import * as util from '../../util';

import { ApiEntity } from '../../types';
import { EntityWrapper } from '../entity-wrapper';

describe('EntityWrapper', () => {
  let entity: ApiEntity;
  let wrapped: EntityWrapper;
  beforeEach(() => {
    entity = {
      _id: 'entity1',
      data: {
        title: util.makeField('My Title', 'text', 'title', 'Title'),
        active: util.makeField(false, 'boolean', 'active', 'Active'),
        count: util.makeField('100', 'number', 'count', 'Count'),
        address: util.makeField({
          city: util.makeField('Nairobi', 'text', 'city', 'City'),
          street: util.makeField('Some street', 'text', 'street', 'Street'),
          landmarks: util.makeField([
            { type: 'text', value: 'Near historical monument' }
          ], 'list', 'landmarks', 'Landmarks')
        }, 'object', 'address', 'Address'),
        lastSeen: util.makeField('2018-11-22T09:19:33.885Z', 'datetime', 'lastSeen', 'Last seen'),
        image: util.makeField({
          ref: 'file1',
          mimeType: 'image/jpeg',
          downloadUrl: 'download'
        }, 'file', 'image', 'Image'),
        misc: util.makeField([
          {
            type: 'file',
            value: {
              ref: 'file2',
              mimeType: 'application/pdf',
              downloadUrl: 'download2'
            }
          },
          { type: 'number', value: '223.45' },
          { type: 'object', value: {
            foo: util.makeField('bar', 'text', 'foo', 'Foo')
          }}
        ], 'list', 'misc', 'Misc')
      },
      createdAt: '2017-02-22T09:19:33.885Z',
      updatedAt: '2018-10-22T09:19:33.885Z',
      createdBy: {
        _id: 'user1',
        type: 'user'
      },
      tags: ['posts', 'stuff']
    };
    wrapped = new EntityWrapper(entity);
  });

  describe('top-level entity metadata', () => {
    it('should offer direct access _id, tags, createdBy properties', () => {
      expect(wrapped._id).toEqual(entity._id);
      expect(wrapped.tags).toEqual(entity.tags);
      expect(wrapped.createdBy).toEqual(entity.createdBy);
    });
    it('should return timestamps properties converted to Date objects', () => {
      expect(wrapped.createdAt).toEqual(new Date(entity.createdAt));
      expect(wrapped.updatedAt).toEqual(new Date(entity.updatedAt));
    });
  });

  describe('rawEntity', () => {
    it('should return the wrapped raw entity', () => {
      expect(wrapped.rawEntity).toEqual(entity);
    });
  });

  describe('value', () => {
    it('should return the plain value at the specified key path', () => {
      expect(wrapped.value('title')).toBe(entity.data.title.value);
    });
    it('should return the plain object at the specified key path', () => {
      expect(wrapped.value('address')).toMatchSnapshot();
    });
    it('should return the plain array at the specified key path', () => {
      expect(wrapped.value('misc')).toMatchSnapshot();
    });
    it('should call util.walkEntityPath to get value at key path', () => {
      const spy = jest.spyOn(util, 'walkEntityPath');
      wrapped.value(['misc', 2]);
      expect(spy.mock.calls).toMatchSnapshot();
    });
    describe('when string path is provided', () => {
      it('should split string path to array', () => {
        const spy = jest.spyOn(util, 'walkEntityPath');
        const stringPath = 'address.city';
        const arrayPath = ['address', 'city'];
        wrapped.value(stringPath);
        expect(spy).toHaveBeenCalledWith(entity, arrayPath);
      });
    });

    describe('dotted field paths', () => {
      it('should support dot notation for nested fields', () => {
        expect(wrapped.value('address.landmarks')).toMatchSnapshot();
      });
    });
    describe('array field paths', () => {
      it('should support array-based paths', () => {
        expect(wrapped.value(['address', 'city'])).toEqual('Nairobi');
      });
      it('should support singleton arrays', () => {
        expect(wrapped.value(['title'])).toEqual(entity.data.title.value);
      });
    });
    describe('when the path is empty string', () => {
      it('should return entity data as plain dehydrated object', () => {
        expect(wrapped.value('')).toMatchSnapshot();
      });
    });
    describe('when the path is empty array', () => {
      it('should return entity data as plain dehydrated object', () => {
        expect(wrapped.value([])).toMatchSnapshot();
      });
    });
    describe('when the path stops at an array index', () => {
      it('should return historical monument item', () => {
        expect(wrapped.value('address.landmarks.0')).toMatchSnapshot();
      });
      it('should return historical monument item when using array paths', () => {
        expect(wrapped.value(['address', 'landmarks', '0'])).toMatchSnapshot();
      });
      it('should return historical monument item when using array paths with integer index', () => {
        expect(wrapped.value(['address', 'landmarks', 0])).toMatchSnapshot();
      });
    });
    describe('when the path goes through an array index', () => {
      it('should return bar', () => {
        expect(wrapped.value('misc.2.foo')).toMatchSnapshot();
      });
      it('should return bar when using array paths', () => {
        expect(wrapped.value(['misc', '2', 'foo'])).toMatchSnapshot();
      });
      it('should return bar when using array paths with integer index', () => {
        expect(wrapped.value(['misc', 2, 'foo'])).toMatchSnapshot();
      });
    });
  });
});