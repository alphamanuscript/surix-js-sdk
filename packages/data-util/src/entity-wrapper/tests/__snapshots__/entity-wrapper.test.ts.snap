// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityWrapper value dotted field paths should support dot notation for nested fields 1`] = `
Array [
  "Near historical monument",
]
`;

exports[`EntityWrapper value should call util.walkEntityPath to get value at key path 1`] = `
Array [
  Array [
    Object {
      "_id": "entity1",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "user1",
        "type": "user",
      },
      "data": Object {
        "active": Object {
          "label": "Active",
          "name": "active",
          "type": "boolean",
          "value": false,
        },
        "address": Object {
          "label": "Address",
          "name": "address",
          "type": "object",
          "value": Object {
            "city": Object {
              "label": "City",
              "name": "city",
              "type": "text",
              "value": "Nairobi",
            },
            "landmarks": Object {
              "label": "Landmarks",
              "name": "landmarks",
              "type": "list",
              "value": Array [
                Object {
                  "type": "text",
                  "value": "Near historical monument",
                },
              ],
            },
            "street": Object {
              "label": "Street",
              "name": "street",
              "type": "text",
              "value": "Some street",
            },
          },
        },
        "count": Object {
          "label": "Count",
          "name": "count",
          "type": "number",
          "value": "100",
        },
        "image": Object {
          "label": "Image",
          "name": "image",
          "type": "file",
          "value": Object {
            "downloadUrl": "download",
            "mimeType": "image/jpeg",
            "ref": "file1",
          },
        },
        "lastSeen": Object {
          "label": "Last seen",
          "name": "lastSeen",
          "type": "datetime",
          "value": "2018-11-22T09:19:33.885Z",
        },
        "misc": Object {
          "label": "Misc",
          "name": "misc",
          "type": "list",
          "value": Array [
            Object {
              "type": "file",
              "value": Object {
                "downloadUrl": "download2",
                "mimeType": "application/pdf",
                "ref": "file2",
              },
            },
            Object {
              "type": "number",
              "value": "223.45",
            },
            Object {
              "type": "object",
              "value": Object {
                "foo": Object {
                  "label": "Foo",
                  "name": "foo",
                  "type": "text",
                  "value": "bar",
                },
              },
            },
          ],
        },
        "title": Object {
          "label": "Title",
          "name": "title",
          "type": "text",
          "value": "My Title",
        },
      },
      "tags": Array [
        "posts",
        "stuff",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
    },
    Array [
      "misc",
      2,
    ],
  ],
]
`;

exports[`EntityWrapper value should return the plain array at the specified key path 1`] = `
Array [
  Object {
    "downloadUrl": "download2",
    "mimeType": "application/pdf",
    "ref": "file2",
  },
  223.45,
  Object {
    "foo": "bar",
  },
]
`;

exports[`EntityWrapper value should return the plain object at the specified key path 1`] = `
Object {
  "city": "Nairobi",
  "landmarks": Array [
    "Near historical monument",
  ],
  "street": "Some street",
}
`;

exports[`EntityWrapper value when the path goes through an array index should return bar 1`] = `"bar"`;

exports[`EntityWrapper value when the path goes through an array index should return bar when using array paths 1`] = `"bar"`;

exports[`EntityWrapper value when the path goes through an array index should return bar when using array paths with integer index 1`] = `"bar"`;

exports[`EntityWrapper value when the path is empty array should return entity data as plain dehydrated object 1`] = `
Object {
  "active": false,
  "address": Object {
    "city": "Nairobi",
    "landmarks": Array [
      "Near historical monument",
    ],
    "street": "Some street",
  },
  "count": 100,
  "image": Object {
    "downloadUrl": "download",
    "mimeType": "image/jpeg",
    "ref": "file1",
  },
  "lastSeen": 2018-11-22T09:19:33.885Z,
  "misc": Array [
    Object {
      "downloadUrl": "download2",
      "mimeType": "application/pdf",
      "ref": "file2",
    },
    223.45,
    Object {
      "foo": "bar",
    },
  ],
  "title": "My Title",
}
`;

exports[`EntityWrapper value when the path is empty string should return entity data as plain dehydrated object 1`] = `undefined`;

exports[`EntityWrapper value when the path stops at an array index should return historical monument item 1`] = `"Near historical monument"`;

exports[`EntityWrapper value when the path stops at an array index should return historical monument item when using array paths 1`] = `"Near historical monument"`;

exports[`EntityWrapper value when the path stops at an array index should return historical monument item when using array paths with integer index 1`] = `"Near historical monument"`;
